client/ayon_core/host/interfaces/workfiles.py

```diff
@@ -0,0 +1,1214 @@
+
+        Not all hosts can know if the current scene is saved because the API
+            of DCC does not support it.
+
+        Returns:
+            Optional[bool]: True if scene is saved and False if has unsaved
+                modifications. None if can't tell if workfiles has
+                modifications.
+
+        """
+        return None
+
+    def get_workfile_extensions(self) -> list[str]:
+        """Extensions that can be used to save the workfile to.
+
+        Notes:
+            Method may not be used if 'list_workfiles' and
+                'list_published_workfiles' are re-implemented with different
+                logic.
+
+        Returns:
+            list[str]: List of extensions that can be used for saving.
+
+        """
+        return []
+
+    def save_workfile_with_context(
+        self,
+        filepath: str,
+        folder_entity: dict[str, Any],
+        task_entity: dict[str, Any],
+        *,
+        version: Optional[int] = None,
+        comment: Optional[str] = None,
+        description: Optional[str] = None,
+        rootless_path: Optional[str] = None,
+        workfile_entities: Optional[list[dict[str, Any]]] = None,
+        project_settings: Optional[dict[str, Any]] = None,
+        project_entity: Optional[dict[str, Any]] = None,
+        anatomy: Optional["Anatomy"] = None,
+    ):
+        """Save the current workfile with context.
+
+        Arguments 'rootless_path', 'workfile_entities', 'project_entity'
+            and 'anatomy' can be filled to enhance efficiency if you already
+            have access to the values.
+
+        Argument 'project_settings' is used to calculate 'rootless_path'
+            if it is not provided.
+
+        Notes:
+            Should this method care about context change?
+
+        Args:
+            filepath (str): Where the current scene should be saved.
+            folder_entity (dict[str, Any]): Folder entity.
+            task_entity (dict[str, Any]): Task entity.
+            version (Optional[int]): Version of the workfile. Information
+                for workfile entity. Recommended to fill.
+            comment (Optional[str]): Comment for the workfile.
+                Usually used in the filename template.
+            description (Optional[str]): Artist note for the workfile entity.
+            rootless_path (Optional[str]): Prepared rootless path of
+                the workfile.
+            workfile_entities (Optional[list[dict[str, Any]]]): Pre-fetched
+                workfile entities for the task.
+            project_settings (Optional[dict[str, Any]]): Project settings.
+            project_entity (Optional[dict[str, Any]]): Project entity.
+            anatomy (Optional[Anatomy]): Project anatomy.
+
+        """
+        self._before_workfile_save(
+            filepath,
+            folder_entity,
+            task_entity,
+        )
+        project_name = self.get_current_project_name()
```

Comment:
This is calling `self.get_current_project_name()` that is not implemented in `IWorkfileHost` so it assumes that this interface is implementing something else too that is not defined here?

Reply:
It is expected the Iterface is combined with `HostBase`. Don't know how to achieve that to not break hosts because of how python implement abstractions.

Reply:
I really don't want to do that. This is interface that is pre-implementing stuff because we're adding ad-hoc functionality to base interface that should be in ideal case implemented by the host itself, or we should have prepared pre-implemented interface for that elsewhere -> but that would still use the methods that are defined on `HostBase`. Adding `HostBase` to inheritance would probably break more things based on how I know python class inheritance works.

I am afraid of what adding this might cause, breaking addons just because "this is coded by a book" doesn't sound good to me.